tango/integrations/torch/eval.py

```diff
@@ -0,0 +1,187 @@
+            Should contain the test data.
+        dataloader : :class:`DataLoader`
+            The data loader that generates test batches. The batches should be :class:`dict`
+            objects.
+        test_split : :class:`str`, optional
+            The name of the data split used for evaluation in the ``dataset_dict``.
+            Default is "test".
+        seed : :class:`int`, optional
+            Used to set the RNG states at the beginning of training.
+        eval_steps : :class:`int`, optional
+            The number of steps to evaluate for. If not specified evaluation will
+            stop after a complete iteration through the ``dataloader``.
+        log_every : :class:`int`, optional
+            Log every this many steps. Default is ``1``.
+        metric_names : ``Sequence[str]``, optional
+            The names of the metrics to track and aggregate. Default is ``("loss",)``.
+        auto_aggregate_metrics : :class:`bool`, optional
+            If ``True`` (the default), the metrics will be averaged across batches.
+            This may not be the correct behavior for some metrics (such as F1),
+            in which you should set this to ``False`` and handle the aggregation
+            internally in your model or with an :class:`EvalCallback`
+            (using :meth:`EvalCallback.post_batch()`).
+        callbacks : ``List[TrainCallback]``
+            A list of :class:`EvalCallback`.
+
+        """
+        set_seed_all(seed)
+
+        check_dataset(dataset_dict, test_split)
+
+        # Resolve device.
+        device = resolve_device()
+
+        # Prep model.
+        model = model.eval().to(device)
+
+        # Construct dataloader.
+        dataloader: DataLoader = dataloader.construct(dataset=dataset_dict[test_split])
+
+        steps: int
+        try:
+            dataloader_len = len(dataloader)
+            steps = dataloader_len if eval_steps is None else min(dataloader_len, eval_steps)
+        except TypeError:
+            if eval_steps is None:
+                raise ConfigurationError(
+                    "You must set 'eval_steps' for streaming/iterable datasets"
+                )
+            else:
+                steps = eval_steps
+
+        # Initialize callbacks.
+        callbacks: List[EvalCallback] = [
+            callback.construct(
+                step_id=self.unique_id,
+                work_dir=self.work_dir,
+                model=model,
+                dataset_dict=dataset_dict,
+                dataloader=dataloader,
+            )
+            for callback in (callbacks or [])
+        ]
+        for callback in callbacks:
+            callback.pre_eval_loop()
+
+        eval_batches = enumerate(islice(dataloader, steps))
+
+        running_metrics: Dict[str, float] = defaultdict(float)
+        aggregated_metrics: Dict[str, float] = {}
+
+        with Tqdm.tqdm(eval_batches, desc="Evaluating", total=steps) as batch_iter:
```

Comment:
You could use `threaded_generator` here. It might make things faster, depending on how long the data loader takes. The problem is, when something goes wrong, it's hard to debug the stack traces you get. Maybe for these kinds of steps we should err on the side of debuggability.